packages/firestore/src/util/async_queue.ts

```diff
@@ -20,30 +20,103 @@ import { AnyDuringMigration, AnyJs } from './misc';
 import { Deferred } from './promise';
 import { Code, FirestoreError } from './error';
 
-type DelayedOperation<T> = {
+/** External Result of scheduling a delayed operation. */
+export interface DelayedOperationResult<T> {
+  /** A promise that will resolve once the operation has been run. */
+  promise: Promise<T>;
+
+  /**
+   * Prevents the operation from running and rejects the promise with a
+   * Code.CANCELLED error.
+   */
+  cancel(): void;
+}
+
+/**
+ * Represents an operation scheduled to be run in the future.
+ *
+ * Created via DelayedOperation.createAndSchedule().
+ * Supports cancellation (via cancel()) and early execution (via scheduleNow()).
+ */
+class DelayedOperation<T> implements DelayedOperationResult<T> {
   // tslint:disable-next-line:no-any Accept any return type from setTimeout().
-  handle: any;
-  op: () => Promise<T>;
-  deferred: Deferred<T>;
-};
+  private timerHandle: any;
+  private readonly deferred = new Deferred<T>();
+  /** true if the operation has not been executed or cancelled yet. */
+  private pending = true;
+
+  static createAndSchedule<T>(
```

Comment:
This seems unnecessarily circuitous:

  * scheduleWithDelay calls
  * createAndSchedule
  * which after a while calls scheduleNow(asyncQueue)
  * which calls asyncQueue.schedule

It mixes state in closures and in the classes in ways that are pretty confusing.
  
Relatedly you have `createAndSchedule` and `scheduleNow` as two vastly different operations which just from their names I would surmise as having similar behavior.

I'd suggest the following changes:
  * Make the asyncQueue explicitly a member of DelayedOperation
  * Rename scheduleNow to conflict less with createAndSchedule. Maybe submit or resume?
  * I'd consider just splitting creation of the op from the deferred scheduling so that you can get rid of createAndSchedule altogether. New up your DelayedOperation and then call delay on it.

Reply:
* Made `asyncQueue` a member.
* Renamed `scheduleNow()` to a private `handleDelayElapsed()`, and added a new `skipDelay()` method.
* I'm inclined to keep `createAndSchedule()` as it prevents the DelayedOperation from being in a created-but-not-yet-scheduled state.  As-is, `timerHandle` is initially defined and then becomes null once the op has been executed or canceled.  I'd rather not have it go null=>defined=>null with null meaning two different things.  The other option is to have the constructor automatically schedule the op, but I don't really like constructors with side-effects.

By the way, part of the original naming problem is that `AsyncQueue.schedule()` is a bad name. It should be `AsyncQueue.enqueue()`, and then `scheduleNow()` could be `enqueueNow()`. But I don't want to do that in this PR.